speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.cpp

```diff
@@ -98,64 +121,307 @@ void FileDataStorage::Shutdown()
+{
+    p_total += p_value;
+}
+
+/// @brief Calculates the average of the past time steps for a variable and resets it
+/// @param p_total The current total of the variable
+/// @return The average of the past time steps for a variable
+float FileDataStorage::GetAverage(float& p_total) const
+{
+    float average = p_total / static_cast<float>(m_compressionRate);
+    p_total = 0;
+    return average;
+}
+
+/// @brief Add the new value to the array in the correct compression step
+/// @param p_values Array with values from the current compression step
+/// @param p_value The new value of this time step for the variable
+/// @param p_compressionStep The current compression step
+template <typename TNumber>
+void FileDataStorage::AddToArray(TNumber p_values[], TNumber p_value, unsigned long p_compressionStep) const
+{
+    int placeInArray = static_cast<int>(p_compressionStep % static_cast<unsigned long>(m_compressionRate));
+    p_values[placeInArray] = p_value;
+}
+
+/// @brief Get the median of the current compression step using Randomized QuickSelect
+/// @param p_values Array with values from the current compression step
+/// @return The median of the past time steps for a variable
+float FileDataStorage::GetMedian(float* p_values)
+{
+    float startPartition = -1, median = -1;
+    GetMedianUtil(p_values, 0, m_compressionRate - 1, static_cast<int>(m_compressionRate / 2), startPartition, median);
+    return median;
+}
+
+/// @brief Utility function to recursively get the median
+/// @param p_values Array with values from the current compression step
+/// @param p_start start of the current section of the array
+/// @param p_end end of the current section of the array
+/// @param p_middle middle of the current section of the array
+/// @param p_a number before the median of p_values
+/// @param p_b median of p_values
+void FileDataStorage::GetMedianUtil(float* p_values, int p_start, int p_end, int p_middle, float& p_startPartition, float& p_endPartition)
+{
+    if (p_start <= p_end)
     {
-        WRITE_STRING(m_outputStream, "LightsDecision");
-        WRITE_VAR(m_outputStream, p_decisions.GetLights());
+        int partitionIndex = RandomPartition(p_values, p_start, p_end);
+
+        if (partitionIndex == p_middle)
+        {
+            p_endPartition = p_values[partitionIndex];
+            if (p_startPartition != -1)
+            {
+                return;
+            }
+        }
+        else if (partitionIndex == p_middle - 1)
+        {
+            p_startPartition = p_values[partitionIndex];
+            if (p_endPartition != -1)
+            {
+                return;
+            }
+        }
+
+        if (partitionIndex >= p_middle)
+        {
+            GetMedianUtil(p_values, p_start, partitionIndex - 1, p_middle, p_startPartition, p_endPartition);
+        }
+        else
+        {
+            GetMedianUtil(p_values, partitionIndex + 1, p_end, p_middle, p_startPartition, p_endPartition);
+        }
```

Comment:
else is not needed here, it can be written as
```suggestion
        if (partitionIndex >= p_middle)
        {
            GetMedianUtil(p_values, p_start, partitionIndex - 1, p_middle, p_startPartition, p_endPartition);
            return;
        }
        GetMedianUtil(p_values, partitionIndex + 1, p_end, p_middle, p_startPartition, p_endPartition);
```